src/api/tap.rs

```diff
@@ -0,0 +1,322 @@
+
+/// Clients can supply glob patterns to find matched topology components.
+trait GlobMatcher<T> {
+    fn matches_glob(&self, rhs: T) -> bool;
+}
+
+impl GlobMatcher<&str> for String {
+    fn matches_glob(&self, rhs: &str) -> bool {
+        match glob::Pattern::new(self) {
+            Ok(pattern) => pattern.matches(rhs),
+            _ => false,
+        }
+    }
+}
+
+/// A tap notification signals whether a pattern matches a component.
+pub enum TapNotification {
+    Matched,
+    NotMatched,
+}
+
+/// A tap payload can either contain a log event or a notification that's intended
+/// to be communicated back to the client to alert them about the status of the tap request.
+pub enum TapPayload {
+    LogEvent(String, LogEvent),
+    Notification(String, TapNotification),
+}
+
+impl TapPayload {
+    pub fn matched(input_name: &str) -> Self {
+        Self::Notification(input_name.to_string(), TapNotification::Matched)
+    }
+
+    pub fn not_matched(input_name: &str) -> Self {
+        Self::Notification(input_name.to_string(), TapNotification::NotMatched)
+    }
+}
+
+/// A tap sink spawns a process for listening for topology changes. If topology changes,
+/// sinks are rewired to accommodate matched/unmatched patterns.
+pub struct TapSink {
+    _shutdown: ShutdownTx,
+}
+
+impl TapSink {
+    /// Creates a new tap sink, and spawns a handler for watching for topology changes
+    /// and a separate inner handler for events. Uses a oneshot channel to trigger shutdown
+    /// of handlers when the `TapSink` drops out of scope.
+    pub fn new(watch_rx: WatchRx, tap_tx: TapSender, patterns: &[String]) -> Self {
+        let (_shutdown, shutdown_rx) = oneshot::channel();
+
+        tokio::spawn(tap_handler(
+            patterns.iter().cloned().collect(),
+            tap_tx,
+            watch_rx,
+            shutdown_rx,
+        ));
+
+        Self { _shutdown }
+    }
+}
+
+/// Sends a 'matched' tap payload.
+async fn send_matched(mut tx: TapSender, pattern: &str) -> Result<(), SendError<TapPayload>> {
+    debug!(message = "Sending matched notification.", pattern = ?pattern);
+    tx.send(TapPayload::matched(pattern)).await
+}
+
+/// Sends a 'not matched' tap payload.
+async fn send_not_matched(mut tx: TapSender, pattern: &str) -> Result<(), SendError<TapPayload>> {
+    debug!(message = "Sending not matched notification.", pattern = ?pattern);
+    tx.send(TapPayload::not_matched(pattern)).await
+}
+
+/// Makes a `RouterSink` that relays `LogEvent` as `TapPayload::LogEvent` to a client.
+fn make_router(mut tx: TapSender, component_name: &str) -> fanout::RouterSink {
+    let (event_tx, mut event_rx) = futures_mpsc::unbounded();
+    let component_name = component_name.to_string();
+
+    tokio::spawn(async move {
+        debug!(message = "Spawned event handler.", component_name = ?component_name);
+
+        while let Some(ev) = event_rx.next().await {
+            if let Event::Log(ev) = ev {
+                let _ = tx
```

Comment:
Should we handle send errors here? At the least, it'd be nice to log them for debugging purposes. I could also see breaking this while loop.

Reply:
Yes, we definitely should. Even a single error here means something went wrong with the client channel, so there's no need to keep the event handler hanging around. Updated to log + break in 19aa7cd7e1bd9ee4a91bd8140a5e1448e9f8a7b1.